### PR TITLE
Add support for qualifications.

### PR DESCRIPTION
**Pull request by timreyn**
Partial implementation of #2051.  Implements WCIF change made in thewca/wcif#3.

This does the following:

- Adds a column to competition_events for qualification times.
- Adds support in the event editor for selecting qualifications.
- Displays qualifications in the events table.

This does not have any impact on competitors' ability to register for events; I'm saving that for a followup PR.

Screenshots:

image
Setting an average qualification

image
Setting a Top N qualification

image
Events list

image
Text at bottom of events list

### CODE REVIEW

**Comment by timreyn** [comment]
@viroulep would you mind reviewing this PR?

**Comment by Jambrose777** [comment]
Just want to double check the flow of creating this:
- Create competition
- set up Events w/ qualifications
- Go back to the organizer edit view and find the qualification reason before submitting.

If so, that’s okay. It’s a little strange having to refill out the edit page after doing the edit events, but that’s also the flow we have for the main event (which is also part of the reason it is not filled out for a couple competitions).

**Comment by timreyn** [comment]
@Jambrose777 Yes, that's right.

I agree with you that it's a strange order of operations. I think it's okay, though, because qualification times are used very infrequently, and only for major competitions. For the major competitions I've been involved in organizing, we've created the competition weeks before confirming it, and edited the competition form many times. I would not expect to get to the point of confirming the competition and realize that I had missed filling out the qualification reason.

File: WcaOnRails/app/models/competition.rb
Comment (timreyn) [inline WcaOnRails/app/models/competition.rb:197]: This is sort of gross, I'm open to suggestions to make this better.  Without it, one of the tests in competition_spec.rb fails (when changing the id of a competition, line 492).  This is because:

- :uses_qualification? is called in line 185
- which triggers a reload of the competition_events, which is an empty list after changing the ID
- which makes the validation of :main_event_id fail

**Comment by EdHollingdale** [comment]
I noticed that in the screenshots the qualification deadline seemed to be able to be set per-event within the event editor, but was displayed as a single date on the events tab screenshot. What does the event tab do if there are different qualification dates? Should we avoid this by having a "select qualification date" thing pop up in the organiser edit view if qualification standards is selected, or is having different qualification deadlines a desired feature?

**Comment by timreyn** [comment]
@EdHollingdale great question.  The WCIF spec technically allows different deadlines to be set for different events.  I doubt this would be used often, but I could imagine it being used for FMC if using a "top N" qualification type there -- while other events may need a qualification time 1-2 weeks before the registration deadline, to give people time to register after their results are submitted, with a "top N" you may be able to have the qualification deadline a bit closer, since people typically can register regardless of whether they're qualified.

That said, I don't actually expect different qualification dates to be used often, but it would be somewhat more complicated to prohibit it, and also it would be complicated to undo that decision if we decided to change it back to per-event later.

If there are multiple, the text at the bottom looks like this:

image

File: WcaOnRails/app/assets/javascripts/competitions.js
Comment (gregorbg) [inline WcaOnRails/app/assets/javascripts/competitions.js:33]: If the "qualification reason" is not coupled to the checkbox anymore, you should consider changing the description text below the text field to indicate that you need to fill this out if you want to use qualifications in the first place.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [inline WcaOnRails/app/models/competition.rb:1559]: Why is this becoming necessary now? The method is pretty large so probably I'm overlooking something really obvious, but I don't see what this `save` invocation has to do with qualification limits.

File: WcaOnRails/app/models/competition.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition.rb:206]: I have no strong feelings about the `persisted_uses_qualification` hack, as this seems to be a general shortcoming with other properties in the `Competition` model as well. Most notably, the failing `main_event_id` validation that you mention in your comment here also relies on the `with_old_id` mechanism itself.

One thing that makes me wonder however: You are not changing anything about the link between competitions and events itself, you are "only" attaching additional data to the event model. Why does this suddenly break an existing test that used to work before?

**Review by gregorbg — COMMENTED** [review]
Thank you for your contribution :) I have only looked at the backend code and logic so far, as I believe there are other WST members who have more experience in React to provide better reviews on the frontend parts.

File: WcaOnRails/app/webpacker/javascript/edit-events/EditEvents.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/EditEvents.js:158]: `Qualification: ` prefix should be handled via i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:16]: The import paths for these functions have most likely changed

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:45]: Consider moving these util functions to a separate standalone helper file.
Especially `toUTCDate` seems like a useful function to have available in general (or might even exist somewhere already?)

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:39]: i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:41]: i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:43]: i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:47]: Declaring all these components in an in-line fashion within the export seems nasty, is there a specific reason for this?

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:107]: i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js:117]: i18n

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/index.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/index.js:141]: Why is this `UNSAFE` specifically required?

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/index.js
Comment (gregorbg) [inline WcaOnRails/app/webpacker/javascript/edit-events/modals/index.js:145]: Why `UNSAFE`? (same as above)

**Review by gregorbg — CHANGES_REQUESTED** [review]
Hey @timreyn! Are you still interested in working on this? I've provided some high-level comments on the React code now as well.

Please note that between the time when you opened this PR and now, some massive changes to the Webpack structure have occured. Most notably, all components now live under `webpacker/components` (pretty much just a folder rename) and util functions and helpers have been extracted to a separate `webpacker/lib` folder.

If you want to pick this up but you're having trouble rebasing, do not hesitate to reach out and we can arrange a video call or something :)

**Comment by timreyn** [comment]
Sorry for letting this sit for so long! (A year, in fact)

Answered most of your comments. There are a few I haven't gotten to yet (a few of the form "why are you doing X" which I don't remember the answer to). I'll try to get this cleaned up soon.

File: WcaOnRails/app/webpacker/javascript/edit-events/EditEvents.js
Comment (gregorbg) [reply WcaOnRails/app/webpacker/javascript/edit-events/EditEvents.js]: Good point, but that shouldn't keep us from improving localization one step at a time :)
Especially for cases where we already have translations in the existing YML, that's a very low hanging fruit.

File: WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js
Comment (gregorbg) [reply WcaOnRails/app/webpacker/javascript/edit-events/modals/Qualification.js]: Thanks!

File: WcaOnRails/lib/qualification.rb
Comment (gregorbg) [reply WcaOnRails/lib/qualification.rb:58]: Sorry but in my diff it's not fixed. Still stuck with `level` without `@`. Did you forget to check this in or am I overlooking something?

**Review by gregorbg — CHANGES_REQUESTED** [review]
Thanks a lot for picking this up again <3
Our JS linter has a few complaints.

Also, you have committed massive structural changes to `db/structure.sql`. This unfortunately happens when your local SQL server version is on a slightly newer version that has different defaults. Right now, the best solution that we have is to manually revert it to its original state and only insert/commit the lines that should actually be touched by your migration.
I have added comments to the one actual SQL line that can be kept :)

In terms of the `UNSAFE_` stuff in React, I did some digging and it turns out that the `EditEvents` was one of the first ever components that were created before coming up with guidelines and standards. So I'll accept the fact that you're only mimicking our bad work for now, and WST can work on cleaning up these components altogether in a future PR.

File: WcaOnRails/lib/qualification.rb
Comment (gregorbg) [reply WcaOnRails/lib/qualification.rb]: Yup!

File: WcaOnRails/db/structure.sql
Comment (timreyn) [reply WcaOnRails/db/structure.sql:1766]: There's a trailing space here that I seem to have accidentally deleted when regenerating the db structure.

File: WcaOnRails/lib/qualification.rb
Comment (timreyn) [reply WcaOnRails/lib/qualification.rb]: Forgot to check it in :)

**Comment by timreyn** [comment]
How do I run the JS linter? I don't see the output anywhere (only the Ruby tests under "checks").

**Comment by gregorbg** [comment]
Thanks a lot for your prompt fixes, the code is looking good on a "low level" perspective now! The only stuff that is left to figure out are some "high level" perspective questions along the lines of "Why are you doing X?". As you indicated in your previous message, those might take a little time to figure out and I absolutely understand that. So just go through the remaining comment threads that are not yet resolved when you have the time :)

**Comment by gregorbg** [comment]
> How do I run the JS linter? I don't see the output anywhere (only the Ruby tests under "checks").

The linter is hooked into a suite of checks handled via `overcommit`. It's a bit of a ceremony to run this because it has very strong opinions about how settings should be applied and signed, so just look at the "Files changed" view here on GitHub. It conveniently matches the error messages from the CI output to the individual code lines.
If that turns out to be too much work just let me know and I'll run the linting for you

**Comment by timreyn** [comment]
spec/i18n_spec.rb "does not have unused keys" is failing because I defined i18n keys only for use in Javascript. Is there a way to tell that test that those keys are actually used?

**Comment by gregorbg** [comment]
Huh, this might be the first ever use-case of JS-only i18n keys that I'm aware of :D
There is a configuration file at `WcaOnRails/config/i18n-tasks.yml.erb` where you can control which keys are ignored. I suggest you add an entry under `ignore_unused` along with a brief explanatory comment. You can test i18n keys by running `i18n-tasks health` from within the `WcaOnRails` folder (the `i18n-tasks` is made available by `bundle install`)

File: WcaOnRails/app/models/competition_event.rb
Comment (timreyn) [reply WcaOnRails/app/models/competition_event.rb:70]: Without this, hitting "save" does not actually save qualification times -- they're updated on `self` but nothing makes `self` get written to DB. This reproduces with spec/models/competition_wcif_spec.rb:461.

I may not be remembering this fully, but my rough understanding is that all of the other attributes here are standalone DB objects, while qualification is the only one that's saved directly in Events. ActiveRecord is able to figure out that `self.rounds = new_rounds` means that the rounds in `new_rounds` need to be committed, but not that setting `self.qualification` means that `self` needs to be committed.

I don't fundamentally understand why ActiveRecord doesn't write the event, though.

File: WcaOnRails/app/assets/javascripts/competitions.js
Comment (timreyn) [reply WcaOnRails/app/assets/javascripts/competitions.js:33]: Will do. One question, should I update this to be a new i18n key if I'm changing the description text? Or will translators be notified that the English text changed and that they should update the translated copy as well?

**Comment by timreyn** [comment]
> Huh, this might be the first ever use-case of JS-only i18n keys that I'm aware of :D There is a configuration file at `WcaOnRails/config/i18n-tasks.yml.erb` where you can control which keys are ignored. I suggest you add an entry under `ignore_unused` along with a brief explanatory comment. You can test i18n keys by running `i18n-tasks health` from within the `WcaOnRails` folder (the `i18n-tasks` is made available by `bundle install`)

Awesome, thank you

File: WcaOnRails/app/assets/javascripts/competitions.js
Comment (gregorbg) [reply WcaOnRails/app/assets/javascripts/competitions.js:33]: Translators will be notified if the text of an existing key changes.
That's what we keep the `original_hash` in the translated files for ;)

File: WcaOnRails/app/models/competition_event.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition_event.rb:70]: At first glance, I don't think this is something ActiveRecord is ever able to "figure out". Instead, you can see in L63 that the round is written to the database "manually", and so are the extensions. Could you do a similar thing for the qualification, instead of taking the current "brute force save" approach?

Comment (timreyn) [reply WcaOnRails/app/models/competition_event.rb]: Ah -- like this?

File: WcaOnRails/app/assets/javascripts/competitions.js
Comment (timreyn) [reply WcaOnRails/app/assets/javascripts/competitions.js:33]: I don't think this was actually how I wanted this to work -- the text box is only supposed to show up when qualification times are used on the events page. Fixed it to work that way now, rather than updating the text.

File: WcaOnRails/app/models/competition_event.rb
Comment (gregorbg) [reply WcaOnRails/app/models/competition_event.rb]: Yes, although now I'm debating with myself which of the versions is cleaner.
No strong preference here, but perhaps the old `save` approach is a bit more concise despite its generic "sledgehammer" approach...=
I'll leave it up to you to decide whether you want to keep the current `update` or whether to revert to `save`

**Comment by gregorbg** [comment]
Looking good now, as far as I can tell all comments have been thoroughly discussed :)
There are merge conflicts that need to be resolved and then I can think we can consider merging this!

**Comment by timreyn** [comment]
Thanks for the review! Unfortunately I noticed a few more issues I'd like to deal with before merging...

- The WCIF spec for "ranking" doesn't differentiate between ranking by single and average. Competitions currently using it aren't consistently using one or the other -- for FMC in this year's continental championships, NAC and OC are using average, while EC and AC are using single. To address this, I'd like to change the WCIF spec for Qualification to:
```
{
  whenDate: "2022-01-01"
  type: "single" | "average"
  method: "result" | "ranking"
  level: AttemptResult | number
}
```
and this will require changing the class hierarchy in qualification.rb as well. What do you think?
- I'm handling FMC mean qualifications incorrectly -- right now the user enters 30, and it's stored in the WCIF as 30, but that should be 3000. I'll need to update AttemptResultInput to pass the correct value, and then update the code to display it as well.
- Multi Blind qualifications are typically displayed as >= N points, but as implemented it's > N points.

**Comment by timreyn** [comment]
I created https://github.com/thewca/wcif/pull/6 for the WCIF change I mentioned.

**Comment by gregorbg** [comment]
Ad WCIF: That makes a lot of sense, I'm surprised nobody noticed it thus far! The discussion can continue over at your other PR, but in terms of the class hierarchy here won't that 2x2 permutation lead to the Inheritance Diamond of Death?
I mean, you'd probably have to rethink how the entire qualification.rb is written.

Ad FMC: This is actually a quirk in the existing AttemptResultInput component. Right now, if there's any competition using attempt result as an advancement condition in FMC, it will have the wrong value in the DB.
That should be AttemptResultInput lines 79 and 106.

Ad MBLD: Good point, but why are you only making this comment in relation to MBLD? Wouldn't it make sense to consistently enforce all of "time, points, moves" to be either < or <=  ?

**Comment by timreyn** [comment]
> Ad WCIF: That makes a lot of sense, I'm surprised nobody noticed it thus far! The discussion can continue over at your other PR, but in terms of the class hierarchy here won't that 2x2 permutation lead to the Inheritance Diamond of Death? I mean, you'd probably have to rethink how the entire qualification.rb is written.

Unfortunately yes, I will need to rethink that.

> 
> Ad FMC: This is actually a quirk in the existing AttemptResultInput component. Right now, if there's any competition using attempt result as an advancement condition in FMC, it will have the wrong value in the DB. That should be AttemptResultInput lines 79 and 106.

Good point. Fortunately that hasn't happened:

```
mysql> select advancement_condition, count(1) from competition_events join rounds on competition_events.id = rounds.competition_event_id where advancement_condition is not null and event_id="333fm" group by 1;
+-------------------------------+----------+
| advancement_condition         | count(1) |
+-------------------------------+----------+
| {"type":"percent","level":60} |        1 |
| {"type":"percent","level":75} |       64 |
| {"type":"ranking","level":12} |        1 |
| {"type":"ranking","level":24} |        1 |
| {"type":"ranking","level":8}  |        2 |
+-------------------------------+----------+
5 rows in set (0.03 sec)
```

> 
> Ad MBLD: Good point, but why are you only making this comment in relation to MBLD? Wouldn't it make sense to consistently enforce all of "time, points, moves" to be either < or <= ?

A couple of thoughts.
1. The current norm (e.g. https://nac2022.speedcubingcanada.org/en/events) is to use >= for MBLD and < for everything else, and keeping consistency seems like a decent starting point.
2. I have a personal preference for <= over < across the board, though I think that matters more for MBLD than timed events. If the qualification standard is >9 points, I think most people would think of that as "at least 10 points" rather than "more than 9 points", while if the qualification standard is <30.00 seconds, I don't think anyone would think of that as "at most 29.99 seconds".
3. However, on further consideration, changing this for just qualification time would cause inconsistency with the way cutoffs, time limits, and advancement conditions are currently handled, and I think that would be worse than the current situation.

So I propose dropping this, and considering a followup to change the way MBLD cutoffs / qualifications / advancement conditions are displayed in the future.

**Comment by timreyn** [comment]
OK! I did the following:

- Changed the format as described in thewca/wcif#6. I decided to fix the Inheritance Diamond of Death by just having everything in a single Qualification class rather than subclasses.
- Updated AttemptResultInput to handle FMC average AttemptResults correctly, and filed #6999 to apply that to AdvancementCondition too
- Filed #7000 for multi cutoffs + qualifications >= rather than >
- Rebased and handled merge conflicts

So I think pending the WCIF spec change this should be good to go!

**Comment by gregorbg** [comment]
As far as I can tell, with the WCIF change merged this is now good to go...?
I'm asking mainly because my air conditioner broke down last weekend and it's currently 29*C at 70% air humidity in my room which makes it hard to focus and check myself :see_no_evil:

**Comment by timreyn** [comment]
Yep, Jonatan merged it! I pushed one more commit with a few changes based on his review there, once the checks pass this should be good to go.

**Review by gregorbg — APPROVED** [review]

**Comment by DanielEgdal** [comment]
It seems there is a small bug:
If I try adding a qualification requirement and want to use 'By Result' (which is the default) all the other needed fields do not show up. I need to manually change the qualification type to something else, and then back to 'By Result' before the fields show up. 
Picture of how it looks to begin with:
image
Picture of how it looks after changing back and forth: (How I believe it should look from the beginning. 
image

**Pull request merged**
